bundles/org.openhab.binding.matter/src/main/java/org/openhab/binding/matter/internal/MatterHandlerFactory.java

```diff
@@ -0,0 +1,106 @@
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.binding.BaseThingHandlerFactory;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerFactory;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link MatterHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Dan Cunningham - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = { ThingHandlerFactory.class, MatterHandlerFactory.class })
+public class MatterHandlerFactory extends BaseThingHandlerFactory {
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_CONTROLLER, THING_TYPE_NODE,
+            THING_TYPE_ENDPOINT);
+
+    private final MatterStateDescriptionOptionProvider stateDescriptionProvider;
+    private final MatterWebsocketService websocketService;
+    private final MatterChannelTypeProvider channelGroupTypeProvider;
+    private final Set<ControllerHandler> controllers = new HashSet<>();
+
+    @Activate
+    public MatterHandlerFactory(@Reference MatterWebsocketService websocketService,
+            @Reference MatterStateDescriptionOptionProvider stateDescriptionProvider,
+            @Reference MatterChannelTypeProvider channelGroupTypeProvider) {
+        this.websocketService = websocketService;
+        this.stateDescriptionProvider = stateDescriptionProvider;
+        this.channelGroupTypeProvider = channelGroupTypeProvider;
+    }
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        ThingTypeUID baseTypeUID = MatterUIDUtils.baseTypeForThingType(thingTypeUID);
+        return SUPPORTED_THING_TYPES_UIDS.contains(baseTypeUID != null ? baseTypeUID : thingTypeUID);
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (THING_TYPE_CONTROLLER.equals(thingTypeUID)) {
+            ControllerHandler controllerHandler = new ControllerHandler((Bridge) thing, websocketService);
+            controllers.add(controllerHandler);
+            return controllerHandler;
+        }
+
+        ThingTypeUID baseTypeUID = MatterUIDUtils.baseTypeForThingType(thingTypeUID);
+        ThingTypeUID derivedTypeUID = baseTypeUID != null ? baseTypeUID : thingTypeUID;
+
+        if (THING_TYPE_NODE.equals(derivedTypeUID)) {
+            return new NodeHandler((Bridge) thing, this, stateDescriptionProvider, channelGroupTypeProvider);
+        }
+
+        if (THING_TYPE_ENDPOINT.equals(derivedTypeUID)) {
+            return new EndpointHandler(thing, this, stateDescriptionProvider, channelGroupTypeProvider);
+        }
+
+        return null;
+    }
+
+    @Override
+    protected synchronized void removeHandler(ThingHandler thingHandler) {
```

Comment:
Shouldn't you make `controllers` thread safe instead of making the method synchronized ?
If not, I assume you should also make `createHandler` synchronized ?